Sources/Algorithms/AdjacentPairs.swift

```diff
@@ -0,0 +1,271 @@
+extension AdjacentPairsSequence {
+  public struct Iterator {
+    @usableFromInline
+    internal var base: Base.Iterator
+
+    @usableFromInline
+    internal var previousElement: Base.Element?
+
+    @inlinable
+    internal init(base: Base.Iterator) {
+      self.base = base
+    }
+  }
+}
+
+extension AdjacentPairsSequence.Iterator: IteratorProtocol {
+  public typealias Element = (Base.Element, Base.Element)
+
+  @inlinable
+  public mutating func next() -> Element? {
+    if previousElement == nil {
+      previousElement = base.next()
+    }
+
+    guard let previous = previousElement, let next = base.next() else {
+      return nil
+    }
+
+    previousElement = next
+    return (previous, next)
+  }
+}
+
+extension AdjacentPairsSequence: Sequence {
+  @inlinable
+  public func makeIterator() -> Iterator {
+    Iterator(base: base.makeIterator())
+  }
+
+  @inlinable
+  public var underestimatedCount: Int {
+    Swift.max(0, base.underestimatedCount - 1)
+  }
+}
+
+/// A collection of adjacent pairs of elements built from an underlying collection.
+///
+/// In an `AdjacentPairsCollection`, the elements of the *i*th pair are the *i*th
+/// and *(i+1)*th elements of the underlying sequence. The following example
+/// uses the `adjacentPairs()` method to iterate over adjacent pairs of
+/// integers:
+/// ```
+/// for pair in (1...5).adjacentPairs() {
+///     print(pair)
+/// }
+/// // Prints "(1, 2)"
+/// // Prints "(2, 3)"
+/// // Prints "(3, 4)"
+/// // Prints "(4, 5)"
+/// ```
+public struct AdjacentPairsCollection<Base: Collection> {
+  @usableFromInline
+  internal let base: Base
+
+  public let startIndex: Index
+
+  @inlinable
+  internal init(base: Base) {
+    self.base = base
+
+    // Precompute `startIndex` to ensure O(1) behavior,
+    // avoiding indexing past `endIndex`
+    let start = base.startIndex
+    let end = base.endIndex
+    let second = start == end ? start : base.index(after: start)
+    self.startIndex = Index(first: start, second: second)
+  }
+}
+
+extension AdjacentPairsCollection {
+  public typealias Iterator = AdjacentPairsSequence<Base>.Iterator
+
+  @inlinable
+  public func makeIterator() -> Iterator {
+    Iterator(base: base.makeIterator())
+  }
+}
+
+extension AdjacentPairsCollection {
+  public struct Index: Comparable {
+    @usableFromInline
+    internal var first: Base.Index
+    
+    @usableFromInline
+    internal var second: Base.Index
+
+    @inlinable
+    internal init(first: Base.Index, second: Base.Index) {
+      self.first = first
+      self.second = second
+    }
+
+    @inlinable
+    public static func < (lhs: Index, rhs: Index) -> Bool {
+      (lhs.first, lhs.second) < (rhs.first, rhs.second)
+    }
+  }
+}
+
+extension AdjacentPairsCollection: Collection {
+  @inlinable
+  public var endIndex: Index {
+    switch base.endIndex {
+    case startIndex.first, startIndex.second:
+      return startIndex
+    case let end:
+      return Index(first: end, second: end)
+    }
+  }
```

Comment:
I strongly suggest unconditionally representing `endIndex` as `Index(first: base.endIndex, second: base.endIndex)`, and adapting `startIndex` to match this representation in the edge case that `base.count == 1` (instead of the other way around). This is the approach we take in `Windows` as well. Having a consistent representation of `endIndex` often makes it easier to reason about index manipulation logic, and I think you'll find that it will improve some of your code.